CSS Text Underline Position Left and Right Parsing

stwrt opened the pull request:
#### 49c7fca71f8ba7afe8bbf0153bc050b15857fd39
<pre>
CSS Text Underline Position Left and Right Parsing
<a href="https://bugs.webkit.org/show_bug.cgi?id=252770">https://bugs.webkit.org/show_bug.cgi?id=252770</a>

Reviewed by Tim Nguyen.

Add support for parsing CSS Text Underline position left and right values.
This does not add support for multiple values like &quot;under left&quot;.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/parser/CSSParserContext.h:
* Source/WebCore/rendering/TextDecorationPainter.cpp:
(WebCore::TextDecorationPainter::paintBackgroundDecorations):
* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/rendering/style/RenderStyleConstants.h:
* Source/WebCore/rendering/style/StyleRareInheritedData.h:

Canonical link: <a href="https://commits.webkit.org/261052@main">https://commits.webkit.org/261052@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e4311cd3abdd59cafc75e5bff5ce24d5511977ba

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/110376 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/19465 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/42982 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/1762 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/119304 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/114323 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/20906 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10624 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/102621 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/116120 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/15571 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/98762 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/43794 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/97532 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/30427 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/85653 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/99103 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/12130 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/31765 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/100111 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/10193 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/12716 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/8732 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/31244 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/18081 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/51380 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/108134 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7651 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/14568 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/26653 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

nt1m reviewed (commented):
- Seems like there are some tests on WPT for parsing: https://wpt.fyi/results/css/css-text-decor/parsing?view=subtest&q=text-underline-position . Are we now passing them? Do we need to import them?
- from-font/under can be combined with left/right, and this patch doesn't take care of this. Do you plan to take care of this separately?

nt1m commented on Source/WebCore/css/CSSProperties.json:
```diff
@@ -8465,7 +8465,9 @@
-                "from-font"
+                "from-font",
+                "left",
+                "right"
```
I think we should restrict these values behind a setting as long as they're not implemented.

You can use `"settings-flag"` like for `overflow: clip`

stwrt replied:
I just finished implementing the left and right, so this should not be needed.

nt1m commented on Source/WebCore/style/InlineTextBoxStyle.cpp:
```diff
@@ -181,6 +181,26 @@ static float computedUnderlineOffset(const UnderlineOffsetArguments& context)
+        } else
+            computedUnderlineOffset = fontMetrics.ascent() + underlineOffset.lengthOr(gap);
+        break;
+    }
```
Seems suspicious to me that left/right have the same implementation.

stwrt replied:
They should be combined in the latest patch.

nt1m reviewed (commented):
Can we have rendering ref/mismatch tests for the feature?

nt1m commented on Source/WebCore/css/parser/CSSParserContext.h:
```diff
@@ -98,6 +98,7 @@ struct CSSParserContext {
     bool masonryEnabled { false };
     bool cssNestingEnabled { false };
     bool cssPaintingAPIEnabled { false };
+    bool cssTextUnderlinePositionLeftRightEnabled { false };
```
Do we need this?

stwrt replied:
Was getting compile errors without it

nt1m replied:
ah I see overflowClipEnabled is in there, looks fine.

nt1m commented on Source/WebCore/style/InlineTextBoxStyle.cpp:
```diff
@@ -99,7 +99,7 @@ static const RenderElement* enclosingRendererWithTextDecoration(const RenderText
 
         if (!ancestor->isRenderInline()) {
-            // We should always find either the block container or an inline box ancestor inbetween.
+            // We should always find either the block container or an inline box ancestor in between.
```
Can we split changes in this file to a separate patch?

stwrt replied:
Oops

stwrt replied:
Fixed

nt1m commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -1235,6 +1235,20 @@ CSSTextSpacingEnabled:
 
+CSSTextUnderlinePositionLeftRight:
+  type: bool
+  status: unstable
```
```suggestion
  status: testable
```

Otherwise this won't be enabled in tests

nt1m commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -1235,6 +1235,20 @@ CSSTextSpacingEnabled:
+  status: unstable
+  category: css
+  humanReadableName: "CSS text-underline-position: Left Right"
+  humanReadableDescription: "Enable the property text-underline-position Left and Right value support"
```
nit: I'd make left/right lowercase to match css values

stwrt committed:
CSS Text Underline Position Left and Right Parsing
https://bugs.webkit.org/show_bug.cgi?id=252770

Reviewed by Tim Nguyen.

Add support for parsing CSS Text Underline position left and right values.
This does not add support for multiple values like "under left".

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/parser/CSSParserContext.h:
* Source/WebCore/rendering/TextDecorationPainter.cpp:
(WebCore::TextDecorationPainter::paintBackgroundDecorations):
* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
(WebCore::operator<<):
* Source/WebCore/rendering/style/RenderStyleConstants.h:
* Source/WebCore/rendering/style/StyleRareInheritedData.h:

Canonical link: https://commits.webkit.org/261052@main
